core/src/core/transaction.rs

```diff
@@ -525,37 +532,26 @@ pub fn aggregate(transactions: Vec<Transaction>) -> Result<Transaction, Error> {
 		outputs.append(&mut transaction.outputs);
 		kernels.append(&mut transaction.kernels);
 	}
+	let as_block = reward.is_some();
+	if let Some((out, kernel)) = reward {
+		outputs.push(out);
+		kernels.push(kernel);
+	}
 
-	// now sum the kernel_offsets up to give us an aggregate offset for the
-	// transaction
-	let total_kernel_offset = {
-		let secp = static_secp_instance();
-		let secp = secp.lock().unwrap();
-		let mut keys = kernel_offsets
-			.into_iter()
-			.filter(|x| *x != BlindingFactor::zero())
-			.filter_map(|x| x.secret_key(&secp).ok())
-			.collect::<Vec<_>>();
-
-		if keys.is_empty() {
-			BlindingFactor::zero()
-		} else {
-			let sum = secp.blind_sum(keys, vec![])?;
-			BlindingFactor::from_secret_key(sum)
-		}
+	// assemble output commitments set, checking they're all unique
+	let mut out_set = HashSet::new();
+	let all_uniq = {
+		outputs.iter().all(|o| out_set.insert(o.commitment()))
 	};
+	if !all_uniq {
+		return Err(Error::AggregationError);
+	}
 
 	let in_set = inputs
 		.iter()
 		.map(|inp| inp.commitment())
 		.collect::<HashSet<_>>();
 
-	let out_set = outputs
-		.iter()
-		.filter(|out| !out.features.contains(OutputFeatures::COINBASE_OUTPUT))
```

Comment:
We need to be careful with coinbase outputs here?
There's a case here where a block spends a coinbase output and its the same coinbase output for the block reward then we cut-through (bypassing the maturity check).
I don't think we ever want to allow cut-through on coinbase outputs.

Reply:
Yes, I was just thinking about adding an additional transaction-level validation check for this. Going to do it now.

Reply:
On 2nd look, `verify_features` already does that for transactions. Regarding the block case, I don't think said block would be valid. The block `verify_coinbase` will fail as coinbase sum won't work, right? And if a miner decided to broadcast the tx spending his coinbase *before* the block it'd get orphaned, and be invalid after (spending an immature coinbase).

Reply:
Actually this block can't be valid at that point.
The output was cut-through maintaining the sums (inputs vs outputs) but simply removing a kernel will break the summing.
I think.